python/bucketbase/cached_immutable_bucket.py

```diff
@@ -7,28 +7,42 @@
 from bucketbase.errors import DeleteError
 from bucketbase.fs_bucket import AppendOnlyFSBucket
 from bucketbase.ibucket import ShallowListing, IBucket, ObjectStream
+from bucketbase import namedlock
 
 
 class CachedImmutableBucket(IBucket):
-    def __init__(self, cache: IBucket, main: IBucket) -> None:
+    def __init__(self, cache: IBucket, main: IBucket, lock_manager: namedlock.NamedLockManager = None) -> None:
         self._cache = cache
         self._main = main
+        self._lock_manager = lock_manager or namedlock.ThreadLockManager()
 
-    def get_object(self, name: PurePosixPath | str) -> bytes:
+    def _get_with_cache_fallback(self, name:PurePosixPath | str, get_cache_func, fetch_and_cache_func):
+        """Generic method for getting objects with cache fallback"""
+        name_str = str(name)
         try:
-            return self._cache.get_object(name)
+            return get_cache_func(name)
         except FileNotFoundError:
-            _content = self._main.get_object(name)
-            self._cache.put_object(name, _content)
-            return _content
+            with self._lock_manager.get_lock(name_str):
+                try:
+                    return get_cache_func(name)  # Check cache again within lock
+                except FileNotFoundError:
+                    return fetch_and_cache_func(name)
+
+    def get_object(self, name: PurePosixPath | str) -> bytes:
+        def fetch_and_cache(n):
```

Comment:
Code smell. Although this would compile and work in Python, avoid using inner functions in production code